pymbolic/primitives.py

```diff
@@ -924,7 +923,8 @@ def {cls.__name__}_eq(self, other):
 
             return self.__class__ == other.__class__ and {comparison}
 
-        cls.__eq__ = {cls.__name__}_eq
+        if {hash}:
+            cls.__eq__ = {cls.__name__}_eq
```

Comment:
This also guards the `__eq__` override with `hash`. Was playing with `IntG` and it also needs to overwrite equality (seems mostly used in tests?).

Not sure if it's worth adding a separate `eq` flag (to match `dataclass`)?

Reply:
Would these better default to `"__eq__" in cls.__dict__`? It seems like the dataclass decorator also does similar checks to not overwrite the user implementation.

Reply:
- I think a separate `eq` flag makes sense.
- I agree that, by default, user implemenations should not be overwritten.

Reply:
Agreed! Made those changes in 
https://github.com/inducer/pymbolic/compare/3bcf915c6b2a360eb09b98c7aba4d2641cfccf4e..451d51fe3936d7d157de7ab766d7a0d1bda65518